proposals/0033-data-collections.md

```diff
@@ -0,0 +1,244 @@
+
+**Full background:** `slug` was originally introduced alongside the content `id` to have a URL-friendly version of the file path, which can be passed to `getStaticPaths()` for route generation. Data collections are not intended to be used as routes, so we don't want to perpetuate this pattern. `slug` also "slugifies" the file path by removing capitalization and replacing spaces with dashes. If we added this processing to data collection IDs, [collection references](https://github.com/withastro/roadmap/blob/d89e2a4c28379108501aa6bf40d2f8d93d81ad02/proposals/0034-collection-references.md) will be less intuitive to use (i.e. "do I include spaces in the referenced ID here?").
+
+## Internals
+
+Data collections should following the API design of content collections with a stripped-down featureset. To wire up data collections, we will introduce an internal utility that mirrors our `addContentEntryType()` integration function. JSON and YAML will be preconfigured by Astro, as a way to dogfood a generic API for wiring up any data collection type in the future. This example shows an internal implementation for `.json` files:
+
+```ts
+addDataEntryType({
+  extensions: ['.json'],
+  getEntryInfo({ contents, fileUrl }) {
+    // Handle empty JSON files, which cause `JSON.parse` to throw
+    if (contents === undefined || contents === '') return { data: {} };
+
+    const data = JSON.parse(contents);
+
+    if (data == null || typeof data !== 'object')
+      throw new Error(`JSON collection entry ${fileUrl.pathname} must be an object.`);
+
+    return { data };
+  },
+});
+```
+
+Then, we will update our type generator to recognize these data-specific file extensions. This should also raise errors when collections are misconfigured (i.e. `type: 'data'` is missing from the config file) and when a mix of content and data in the same collection is detected.
+
+The `astro:content` runtime module should also be updated to glob these file extensions, and respect the entry ID from the new `getEntry()` utility function. 
+
+# Testing Strategy
+
+- Fixture tests defining data collections, ensuring schemas are respected and query APIs (`getCollection()` and `getEntry()`) return entries of the correct type.
+- Test error states: mixed data / content collections are not allowed, `type: 'data'` is enforced
+
+# Drawbacks
+
+- Data collections could be considered added complexity, when users can create `.md` files storing all data via frontmatter instead. Though true for trivial content, this is restrictive for use cases like i18n lookup files (which are typically JSON).
+
+# Alternatives
+
+## Using multi-file vs. single-file
+
+Early proposals supported single0file data collections. This allows storing _all_ data collection entries as an array in one `.json` file, instead of splitting up entries per-file as we do content collections today. For example, a single `src/content/authors.json` file instead of a few `src/content/authors/[name].json` files.
```

Comment:
```suggestion
Early proposals supported single file data collections. This allows storing _all_ data collection entries as an array in one `.json` file, instead of splitting up entries per-file as we do content collections today. For example, a single `src/content/authors.json` file instead of a few `src/content/authors/[name].json` files.
```